### PR TITLE
Add explorer.autorevealExclude setting

### LINKED ISSUE #87956: Allow to set explorer.autoReveal per specific folder
Issue Type: <b>Feature Request</b>

node_modules will constantly expand itself in the Explorer, especially if your defining definitions.  I typically never want this, I think an exception should be made for showing what file is open if its in node_modules to keep the folder collapsed unless I explicitly expand it.  

The reason it's a problem is because there are so many files in node_modules you have to collapse it to keep developing, as it's completely useless in most cases to have it expanded and it pushes your work files off screen.

Proposition:
- Never expand node_modules implicitly
- If goto definition is used into a node_module, go there but do not older current folder state (if its collapsed keep it collapsed, if it was explicitly expanded by user keep it).
- This behavior applies very specifically to node_modules only.

VS Code version: Code - Insiders 1.42.0-insider (7c0095ee2d064033fc13184127a9adc603454729, 2019-12-20T07:58:23.119Z)
OS version: Linux x64 5.3.0-24-generic

### PR DESCRIPTION
**Pull request by jzyrobert**
This PR fixes #87956

Loosely based on the original PR by @joelday and the `FilesFilter` class, using the suggestion of @JacksonKearl to be in the `ExplorerService` file. 

- Inherits from `files.exclude`
- Defaults to same values as `search.exclude`: `**/node_modules`, `**/bower_components`, `**/*.code-search`
- Calls to `refresh`, `selectResource` and calls from `ExplorerView` are all routed to `ExplorerService.select()`, where the reveal check occurs
- Recursively checks from the file up to the first root child, if any of them match a pattern then the file is not revealed.

Example of glob matching nested folder:
lx4QuzvqPP

Example of using when clause to hide .js files:
RZ9Ji13Xmh

Some things to consider:

- Tests: I wasn't sure if there is support for testing whether the reveal correctly shows or not. `files.exclude` only has tests for the find files function, and not for `FilesFilter`. I suppose you could create a `RevealFilter` manually and check the boolean function, but I didn't see other tests having support for mocking `contextService` etc
  - Edit: Could be added to integration tests?
- Behaviour if folder is already open:
  - Should the file still be revealed?
  - Should it be ignored? (<- current behaviour)
- Behaviour if an open folder is added to the setting:
  - Should the tree close the folder?
  - Should it be left as is? (<- current behaviour)
- Performance:
  - Should the check run on each selection? How much does iterating through the parent tree of an item each time matter (No initial cost, each call only iterates up it's own parent(s))
  - Or should it mirror the `files.exclude` setting, which runs top-down each time files are updated and sets a cached boolean for each folder/file. (Bigger initial cost as has to check all folder/files, no cost after)

### CODE REVIEW

File: src/vs/workbench/contrib/preferences/browser/settingsTreeModels.ts
Comment (codeclown) [inline src/vs/workbench/contrib/preferences/browser/settingsTreeModels.ts:574]: This does not do what it looks like in JS (see the [`in` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in)): `'hey' in ['hey', 'foobar'] === false`

If necessary to change to checking from an array like this (which is contrary to rest of file), should use `['...'].includes(...)`.

Comment (jzyrobert) [reply src/vs/workbench/contrib/preferences/browser/settingsTreeModels.ts:574]: You're right,  I was thinking of python syntax when I wrote this.  Need reverting

Comment (jzyrobert) [reply src/vs/workbench/contrib/preferences/browser/settingsTreeModels.ts]: Fixed

**Comment by JacksonKearl** [comment]
It's not clear do me what the FilesFilter class does here, it seems to duplicate a lot of the existing functionality of the glob module but with more code to maintain, I'd like to see what the smallest possible diff to implement the desired functionality would look like.

**Comment by jzyrobert** [comment]
@JacksonKearl I have implemented it using `ResourceGlobMatcher` instead as you have requested, please take another look.

I'm not sure if its possible the required recursive matching into the matcher, as you cannot import from workbench/contrib

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (JacksonKearl) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:531]: is this export used?

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (JacksonKearl) [inline src/vs/workbench/contrib/files/browser/explorerService.ts:140]: Do we need to refresh?

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (JacksonKearl) [inline src/vs/workbench/contrib/files/browser/explorerService.ts:376]: Couple notes: 
- the name of this isn't very discriptive `shouldAutoRevealItem` might be better?
- walking up the directly tree for every item isn't great, the glob matcher should be able to know that `node_modules/**` matches `node_modules/abc/d` without stepping through every directory?

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (jzyrobert) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:531]: Not anymore if the ResourceGlobMatcher is used instead.

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (jzyrobert) [reply src/vs/workbench/contrib/files/browser/explorerService.ts:140]: If you had a file revealed and then added a filter to it, it should refresh the tree and not reveal it anymore. I believe that is the desired behaviour for settings?

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (jzyrobert) [reply src/vs/workbench/contrib/files/browser/explorerService.ts:376]: I will change the name as you suggested.

The glob matcher can match files if specified, e.g. `**/node_modules/*` would work fine without recursive checking.
The reason for this implementation is that the settings are inherited from the `files.exclude` setting, which allows specifying both folder and file patterns for showing files in the explorer. That setting is passed recursively and only needs to be called once for every tree item.
However, I also don't think the performance impact is too serious, as:

1. Focusing on a file is not a super frequent operation
2. File structures do not tend to be super deep, in the case of `**/node_modules` it would be 4-5 match operations for a source file inside 

In that case, there are two options I would suggest here:

1.  Keep the matching as is, but only check for sibling clauses with the initial item, and do not pass as sibling clause for the parent matching.
2. Remove the recursive matching, and edit the default setting and setting description to say that you must explicitly specify file patterns (`**/node_modules/*`) and not folder patterns (`**/node_modules`)

**Comment by Livan-pro** [comment]
It looks very helpful. @JacksonKearl, what about this?

File: src/vs/workbench/common/resources.ts
Comment (JacksonKearl) [inline src/vs/workbench/common/resources.ts:208]: This cannot return a boolean if the hasSibiling returns a promise.

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (JacksonKearl) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:531]: If the export isnt used then we should just keep it internal

File: src/vs/workbench/common/resources.ts
Comment (JacksonKearl) [inline src/vs/workbench/common/resources.ts:230]: `!!expressionForRoot(resourcePathToMatch, undefined, hasSibling)` will always be true when the result is a promise

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (JacksonKearl) [inline src/vs/workbench/contrib/files/browser/explorerService.ts:391]: Looks like yiou don't actually need the Promise here?

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (JacksonKearl) [inline src/vs/workbench/contrib/files/browser/explorerService.ts:254]: In some contexts the item should be revealed even if it's marked as autoreveal excluded. For instance when running the `revealInExplorer` command (`File: Reveal Active File in Sidebar`)

File: extensions/configuration-editing/src/settingsDocumentHelper.ts
Comment (JacksonKearl) [inline extensions/configuration-editing/src/settingsDocumentHelper.ts:33]: Capital R in Reveal here?

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (JacksonKearl) [reply src/vs/workbench/contrib/files/browser/explorerService.ts:144]: I'd expect the explorer to remain as it is when changing this setting, the setting should only be live to active editor changes.

**Comment by JacksonKearl** [comment]
I wonder how this should interact with recently added file nesting.

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (jzyrobert) [reply src/vs/workbench/contrib/files/browser/explorerService.ts:391]: You're right, I will make it boolean only

File: extensions/configuration-editing/src/settingsDocumentHelper.ts
Comment (jzyrobert) [reply extensions/configuration-editing/src/settingsDocumentHelper.ts:33]: Thanks for spotting that

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (jzyrobert) [reply src/vs/workbench/contrib/files/browser/explorerService.ts]: Do you know why the reveal function accepts `boolean | string`?
It is passed to `view.selectResource()`, but that only ever checks if `reveal === true` and does not make use of the string.

Would it be ok to have the `revealInExplorer` command pass something like `force` for the reveal parameter and have the select function check for it?

**Comment by AmitPr** [comment]
Hi, any update on this?

**Comment by lramos15** [comment]
@jzyrobert Jackson has since left the company so I have taken over reviewing this PR. Please resolve merge conflicts and get CI to pass and then ping me for a review. Hopefully we can get this in for not this release but the next one. Thank you for your contribution.

**Comment by jzyrobert** [comment]
Hi @lramos15 @AmitPr 

I have merged and updated the PR. There was an outstanding question I had with Jackson about still allowing reveals when running the "Reveal current file" command.

The command calls the function the same way as normally selecting on a file, so I changed it to pass in `force` and have the select function check for that.

However the underlying `selectResource` function actually expects only a boolean even if the signature is `bool | string` so I have also made it convert the argument back to `true` if it was `force`

I don't know if there is a better way to detect whether the function is being called from the command without changing the signature of `select` itself to include a third parameter

**Comment by jzyrobert** [comment]
> > However the underlying `selectResource` function actually expects only a boolean even if the signature is `bool | string` so I have also made it convert the argument back to `true` if it was `force`
> 
> Can you provide a code pointer to this? I see `selectResource` accepting a boolean or a string. Just that if it's a string it doesn't really matter because the only string value is `focusNoScroll` and we always focus so false or any string is treated as `false` to the logic in the function. Doesn't mean you couldn't pass `force` in. Would something like `(reveal === true || reveal === 'force')` work here
> 
> https://github.com/microsoft/vscode/blob/f17b33faf21feba942e4663e40226456b081ae5d/src/vs/workbench/contrib/files/browser/views/explorerView.ts#L749

No you're right, that's the line I'm referring to. If you would prefer a `== force` check there I'm happy to change it

**Comment by jzyrobert** [comment]
> > No you're right, that's the line I'm referring to. If you would prefer a `== force` check there I'm happy to change it
> 
> Yes please I think it makes it a bit clearer what's going on if it is `(reveal === true || reveal === force)` since this is another possibility for reveal

Pushed those changes

**Review by riverpaper — APPROVED** [review]

File: src/vs/workbench/contrib/files/browser/explorerService.ts
Comment (lramos15) [reply src/vs/workbench/contrib/files/browser/explorerService.ts]: @jzyrobert Did we settle on anything here I think we only support file patterns for files.exclude so this should mirror that.

**Comment by lramos15** [comment]
@jzyrobert Overall this looks great! I saw in the initial description there was discussions surrounding on how to test this. If you're feeling up for implementing tests that would be great since this is a new untested feature. Unit tests or integration tests are fine, unit tests are a bit faster and we do have mocks for all our services

**Review by riverpaper — APPROVED** [review]

**Comment by TurnipGuy30** [comment]
Sorry for the second review...

**Comment by mindplay-dk** [comment]
As suggested [here](https://github.com/microsoft/vscode/issues/87956#issuecomment-1174716054):

> 💡 How about not auto-expanding things that are in your `.gitignore`?
> 
> That would work for `node_modules` and probably for other similar scanarios - without any manual intervention.
> 
> It's great to have options - but it would be even nicer if (most) users didn't need to solve this problem in the first place.
> 
> Those items are already grayed-out in the tree-view, so the IDE already "knows" - most likely, folders you wish to ignore in terms of source-control are just artifacts that the user wants to ignore most of the time in practice as well.

I think this would make sense as a default?

**Review by riverpaper — DISMISSED** [review]

**Review by gjsjohnmurray — CHANGES_REQUESTED** [review]
I suggest 
- renaming the setting to `explorer.autoReveal.exclude`
- changing `explorer.autoReveal` from a boolean to an enum with values off, on and smart. Existing setting value would need to be migrated, but I think there is a precedent for doing that. 
- only applying exclude when user has set the setting to smart.

This makes the new behaviour opt-in. Otherwise we risk upsetting people who haven't read the release notes and are confused by the sudden change in what they are used to.

**Comment by lramos15** [comment]
> * renaming the setting to `explorer.autoReveal.exclude`

I agree



> * changing `explorer.autoReveal` from a boolean to an enum with values off, on and smart. Existing setting value would need to be migrated, but I think there is a precedent for doing that.

This seems overly complex and confusing to me. I get more issues of people being upset about the `node_modules` folder expanding automatically than not. I would say we're probably safe here, but if we want to not cause any new behavior I would prefer having no defaults for this setting rather than changing `explorer.autoReveal` which is a long standing setting and has been for awhile. Migrating a setting while possible isn't easy and causes quite a bit of headache on the dev side.

**Comment by lramos15** [comment]
Also the when clause use here seems to be a bit strange to me.

You have
`"**/unwanted_when/*.js: {"when": "${basename}.ts}`.

You show the `unwanted_when` folder expanding when you open a `*.ts`, but not when you open a `*.js` file. Wouldn't you expect this to be accomplished just like this `"**/unwanted_when/*.js: true` or I would expect all globs to be folders and then you could do `"**/unwanted_when: {"when": "*.js"}` because then it would be true when it's a JS file meaning we exclude it from auto reveal

**Comment by lramos15** [comment]
@jzyrobert Just wanted to bump this in case you're still interested in contributing, otherwise I can take over this feature :)

**Comment by jzyrobert** [comment]
> @jzyrobert Just wanted to bump this in case you're still interested in contributing, otherwise I can take over this feature :)

Mainly forgot about it after getting stuck with the tests. Still not sure how you would add some for this. I would assume in `vscode/src/vs/workbench/contrib/files/test/browser` would be the closest but there isn't an `TestEplorerService`. Would you be willing to take a go at the tests, or accept this feature without them?

And to address some of the other points:

@gjsjohnmurray 
> I suggest
> 
> * renaming the setting to `explorer.autoReveal.exclude`

I don't think this is possible or allowed because `explorer.autoReveal` is the existing setting. Either settings are named `setting.subsetting1`, `setting.subsetting2` and there is no `setting`, or they are named `setting` and `settingSubsetting` (e.g. `explorer.sortOrder` and `explorer.sortOrderLexicographicOptions`, not `explorer.sortOrder.LexicographicOptions`). I couldn't find an existing setting which has both `setting` and `setting.subsetting`

> Also the when clause use here seems to be a bit strange to me.
> 
> You have `"**/unwanted_when/*.js: {"when": "${basename}.ts"}`.
> 
> You show the `unwanted_when` folder expanding when you open a `*.ts`, but not when you open a `*.js` file. Wouldn't you expect this to be accomplished just like this `"**/unwanted_when/*.js: true` or I would expect all globs to be folders and then you could do `"**/unwanted_when: {"when": "*.js"}` because then it would be true when it's a JS file meaning we exclude it from auto reveal

The behaviour matches exactly how `files.exclude` does it. Globs for both folders and files are allowed. One of the default settings at `vscode/.vscode/settings.json` is
```json
	"files.exclude": {
		"build/**/*.js": {
			"when": "$(basename).ts"
		}
```
and this means that I only want to exclude js files _when_ a ts file with the same name exists in the same folder. If you wanted to exclude all js files it would just be. The alternatives `"**/unwanted_when/*.js: true` you suggest would always exclude all js files, which is not what we want when using the `when` condition.

> @jzyrobert Did we settle on anything here I think we only support file patterns for files.exclude so this should mirror that.

`files.exclude` definitely supports folder patterns, which is stated in the setting description:
```typescript
"Configure [glob patterns](https://code.visualstudio.com/docs/editor/codebasics#_advanced-search-options) for excluding files and folders."
```

and in the default values which includes `**/.git`, `**/.svn` etc which are folders and not files.

@mindplay-dk 
>💡 How about not auto-expanding things that are in your .gitignore?
>
>That would work for node_modules and probably for other similar scanarios - without any manual intervention.
>
>It's great to have options - but it would be even nicer if (most) users didn't need to solve this problem in the first place.
>
>Those items are already grayed-out in the tree-view, so the IDE already "knows" - most likely, folders you wish to ignore in terms of source-control are just artifacts that the user wants to ignore most of the time in practice as well.

I think this should be an opt in (configure it, make it an extra setting, etc) rather than opt-out. The git setting only grays out the folders, but it would confuse users if the reveal didn't behave like expected. E.g. `files.exclude` has an extra setting for this:
```typescript
		'explorer.excludeGitIgnore': {
			type: 'boolean',
			markdownDescription: nls.localize('excludeGitignore', "Controls whether entries in .gitignore should be parsed and excluded from the explorer. Similar to {0}.", '`#files.exclude#`'),
			default: false,
			scope: ConfigurationScope.RESOURCE
		},
```

So if requested, an additional setting like `explorer.autorevealExcludeGitignore` could be added.

**Comment by lramos15** [comment]
I'm a bit hesitant in inheriting from `files.exclude` I quite like that if I do open those files they appear in the explorer so that they can be edited and acted upon. I don't mind sharing the same defaults, but I don't know if I'm comfortable in them always being merged.

File: src/vs/workbench/contrib/files/browser/files.contribution.ts
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/files.contribution.ts:377]: What is code-search and why is it in the default?

File: src/vs/workbench/contrib/files/browser/files.contribution.ts
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/files.contribution.ts:376]: I do not understand how is the inheriting done and how can the user change that?
How can the user discover that if they miss this desciprtion? We do not have inheritence structure in our settings and I suggest that we do not introduce it now (unless I am missing some case)

**Review by isidorn — COMMENTED** [review]
Thank you for this PR. I have left some minor comments. It would be great if @bpasero also gives it a glance for exclude and globs.

**Review by bpasero — DISMISSED** [review]
I do not fully understand the need for glob patterns that check on siblings for this purpose here: imho extending from `files.exclude` does not really make a lot of sense because the explorer will not show anything that is `file.exclude` (at least to my knowledge) and as such will not auto-reveal anyway. And for the setting I would believe simple glob patterns are sufficient and not the ones with sibling support?

**Comment by jzyrobert** [comment]
> I do not fully understand the need for glob patterns that check on siblings for this purpose here: imho extending from `files.exclude` does not really make a lot of sense because the explorer will not show anything that is `file.exclude` (at least to my knowledge) and as such will not auto-reveal anyway. And for the setting I would believe simple glob patterns are sufficient and not the ones with sibling support?
@bpasero 

The explorer will show files in `file.exclude`, for example:
image
The tree shows nothing because I excluded `hiddenFolder`

Then I access a file in there by some way (Ctrl+ P or if it was imported etc):
image
It shows up in the explorer. What it doesn't show is anything else in that folder.

@isidorn makes a point that it wouldn't be obvious to users unless they find the setting description so that can be removed and users can add back what they want.

I can remove the sibling support if required, but wouldn't it be useful to have? It doesn't interfere with the normal glob functionality and can be useful such as wanting to check a compiled JS file but having it revealed if a TS file exists with the same name in that folder.

**Comment by lramos15** [comment]
This looks good to me from the explorer side of things and I'm hoping to take this in either today or tomorrow to get ample testing in insiders.

@bpasero I'll defer to you if we should keep sibling support or not. It does seem like it only is wiring through an existing check via matches so unsure if that still feels like tainting the interface. We can probably gauge via telemetry how often `when` patterns are used and decide whether or not we should implement it in the future or remove it in the future deciding which way we want to go.

**Comment by bpasero** [comment]
Yeah its fine for me, doesn't need to block this PR from merging.

**Review by lramos15 — APPROVED** [review]
This is an amazing contribution, thanks for working with us as we understand how long it takes to get something like this merged in!

**Review by mjbvz — APPROVED** [review]

**Comment by lramos15** [comment]
This will be in tomorrows insiders.

**Pull request merged**
